src/NuGet.Services.AzureSearch/SearchService/SecretRefresher.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using NuGet.Services.AzureSearch.Wrappers;
+using NuGet.Services.KeyVault;
+
+namespace NuGet.Services.AzureSearch.SearchService
+{
+    public class SecretRefresher : ISecretRefresher
+    {
+        private readonly IRefreshSecretReaderFactory _factory;
+        private readonly ISystemTime _systemTime;
+        private readonly IOptionsSnapshot<SearchServiceConfiguration> _options;
+        private readonly ILogger<SecretRefresher> _logger;
+
+        public SecretRefresher(
+            IRefreshSecretReaderFactory factory,
+            ISystemTime systemTime,
+            IOptionsSnapshot<SearchServiceConfiguration> options,
+            ILogger<SecretRefresher> logger)
+        {
+            _factory = factory ?? throw new ArgumentNullException(nameof(factory));
+            _systemTime = systemTime ?? throw new ArgumentNullException(nameof(systemTime));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task RefreshContinuouslyAsync(CancellationToken token)
+        {
+            while (!token.IsCancellationRequested)
+            {
+                _logger.LogInformation("Trying to refresh the secrets.");
+
+                TimeSpan delay;
+                try
+                {
+                    await _factory.RefreshAsync(token);
+                    delay = _options.Value.SecretRefreshFrequency;
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogError(0, ex, "An exception was thrown while refreshing the secrets.");
+
+                    delay = _options.Value.SecretRefreshFailureRetryFrequency;
+                }
+
+                if (token.IsCancellationRequested)
```

Comment:
What do y'all think about recording the "last refresh time" and reporting it on `/search/diag` for monitoring?

Reply:
Considerations:

1. I don't think it's harmful to tell this information publicly. It is _related_ to secrets through and could be used to know that secrets haven't been loaded in a long time.
1. I want to put this on `/search/diag` instead of emitting a metric so that a monitoring runner can report a metric only about instances that are in rotation. This is a lot harder to from metrics emitted from a service.

Reply:
May be report this as "LastServiceRefreshTime" on /search/diag, so its not obvious that this is about _secrets_

Reply:
Open source, but sure sounds fine!